kernel/aster-nix/src/net/socket/unix/stream/listener.rs

```diff
@@ -144,42 +136,58 @@ impl BacklogTable {
 }
 
 struct Backlog {
+    addr: UnixSocketAddrBound,
     pollee: Pollee,
-    backlog: usize,
-    incoming_endpoints: Mutex<VecDeque<Endpoint>>,
+    backlog: AtomicUsize,
+    incoming_conns: Mutex<VecDeque<Connected>>,
 }
 
 impl Backlog {
-    fn new(backlog: usize) -> Self {
+    fn new(addr: UnixSocketAddrBound, backlog: usize) -> Self {
         Self {
+            addr,
             pollee: Pollee::new(IoEvents::empty()),
-            backlog,
-            incoming_endpoints: Mutex::new(VecDeque::with_capacity(backlog)),
+            backlog: AtomicUsize::new(backlog),
+            incoming_conns: Mutex::new(VecDeque::with_capacity(backlog)),
         }
     }
 
-    fn push_incoming(&self, endpoint: Endpoint) -> Result<()> {
-        let mut endpoints = self.incoming_endpoints.lock();
-        if endpoints.len() >= self.backlog {
-            return_errno_with_message!(Errno::ECONNREFUSED, "incoming_endpoints is full");
+    fn addr(&self) -> &UnixSocketAddrBound {
+        &self.addr
+    }
+
+    fn push_incoming(&self, client_addr: Option<UnixSocketAddrBound>) -> Result<Connected> {
+        let (server_conn, client_conn) = Connected::new_pair(Some(self.addr.clone()), client_addr);
+
+        let mut incoming_conns = self.incoming_conns.lock();
+        if incoming_conns.len() >= self.backlog.load(Ordering::Relaxed) {
+            return_errno_with_message!(
+                Errno::ECONNREFUSED,
+                "the pending connection queue on the listening socket is full"
+            );
         }
```

Comment:
The check of line 162-167 should be better put before line 160. So if the check fails, we can avoid creating `Connected`.

Reply:
Perhaps it's better to allocate memory before holding the lock. For example, in the Linux kernel, allocating memory while holding a spin lock implies `GFP_ATOMIC`, which is much stricter than `GFP_KERNEL`.

But in our case it probably doesn't make much sense, at least for now. After all, we already have the state lock at this point. So I changed the code to improve readability. (We can assume that a failed case is rare, so avoiding the creation of `Connected` doesn't really bring much benefit).